cdap-app-templates/cdap-etl/cdap-data-streams-base/src/test/java/io/cdap/cdap/datastreams/DataStreamsTest.java

```diff
@@ -1436,4 +1437,106 @@ private long getMetric(ApplicationId appId, String metric) {
                                                Constants.Metrics.Tag.SPARK, DataStreamsSparkLauncher.NAME);
     return getMetricsManager().getTotalMetric(tags, "user." + metric);
   }
+
+  @Test
+  /**
+   * Uses {@link RecoveringTransform} plugin to test retry. Default retry settings are used.
+   */
+  public void testRetryOnException() throws Exception {
+    Schema schema = Schema.recordOf(
+      "retry_test",
+      Schema.Field.of("id", Schema.of(Schema.Type.STRING)),
+      Schema.Field.of("name", Schema.of(Schema.Type.STRING))
+    );
+    List<StructuredRecord> input = new ArrayList<>();
+    StructuredRecord samuelRecord = StructuredRecord.builder(schema).set("id", "123").set("name", "samuel").build();
+    StructuredRecord jacksonRecord = StructuredRecord.builder(schema).set("id", "456").set("name", "jackson").build();
+    input.add(samuelRecord);
+    input.add(jacksonRecord);
+
+    DataStreamsConfig etlConfig = DataStreamsConfig.builder()
+      .addStage(new ETLStage("retry_source", MockSource.getPlugin(schema, input)))
+      .addStage(new ETLStage("retry_sink", MockSink.getPlugin("${retry_output_macro}")))
+      .addStage(new ETLStage("retry_transform", RecoveringTransform.getPlugin()))
+      .addConnection("retry_source", "retry_transform")
+      .addConnection("retry_transform", "retry_sink")
+      .setBatchInterval("1s")
+      .build();
+
+    ApplicationId appId = NamespaceId.DEFAULT.app("simpleRetryApp");
+    AppRequest<DataStreamsConfig> appRequest = new AppRequest<>(APP_ARTIFACT, etlConfig);
+    ApplicationManager appManager = deployApplication(appId, appRequest);
+
+    Set<StructuredRecord> expected = new HashSet<>();
+    expected.add(samuelRecord);
+    expected.add(jacksonRecord);
+
+    SparkManager sparkManager = appManager.getSparkManager(DataStreamsSparkLauncher.NAME);
+    String outputName = "retry_output";
+    sparkManager.start(ImmutableMap.of("retry_output_macro", outputName));
+    sparkManager.waitForRun(ProgramRunStatus.RUNNING, 10, TimeUnit.SECONDS);
+
+    // since dataset name is a macro, the dataset isn't created until it is needed. Wait for it to exist
+    Tasks.waitFor(true, () -> getDataset(outputName).get() != null, 3, TimeUnit.MINUTES);
+
+    DataSetManager<Table> outputManager = getDataset(outputName);
+    Tasks.waitFor(
+      true,
+      () -> {
+        outputManager.flush();
+        Set<StructuredRecord> outputRecords = new HashSet<>(MockSink.readOutput(outputManager));
+        return expected.equals(outputRecords);
+      },
+      3,
+      TimeUnit.MINUTES);
+
+    MockSink.clear(outputManager);
+    sparkManager.stop();
+    sparkManager.waitForStopped(10, TimeUnit.SECONDS);
+    RecoveringTransform.reset();
```

Comment:
stateful things like this need to be in a try-finally (or in an `@After` method) to ensure the tests are isolated from each other.

Reply:
Moved to finally block.